linux/src/threadpool_linux.c

```diff
@@ -544,65 +557,75 @@ int threadpool_timer_start(THANDLE(THREADPOOL) threadpool, uint32_t start_delay_
         timer_handle == NULL
         )
     {
-        LogError("Invalid args: THANDLE(THREADPOOL) threadpool = %p, uint32_t start_delay_ms = %" PRIu32 ", uint32_t timer_period_ms = %" PRIu32 ", THREADPOOL_WORK_FUNCTION work_function = %p, void* work_function_context = %p, TIMER_INSTANCE_HANDLE* timer_handle = %p",
+        LogError("Invalid args: THANDLE(THREADPOOL) threadpool = %p, uint32_t start_delay_ms = %" PRIu32 ", uint32_t timer_period_ms = %" PRIu32 ", THREADPOOL_WORK_FUNCTION work_function = %p, void* work_function_context = %p, TIMER ** timer_handle = %p",
             threadpool, start_delay_ms, timer_period_ms, work_function, work_function_ctx, timer_handle);
         result = MU_FAILURE;
     }
     else
     {
-        /* Codes_SRS_THREADPOOL_LINUX_07_058: [ threadpool_timer_start shall allocate a context for the timer being started and store work_function and work_function_ctx in it. ]*/
-        TIMER_INSTANCE* timer_instance = malloc(sizeof(TIMER_INSTANCE));
-        if (timer_instance == NULL)
+        /* Codes_SRS_THREADPOOL_LINUX_07_058: [ threadpool_timer_start shall allocate memory for THANDLE(TIMER), passing threadpool_timer_dispose as dispose function, and store work_function and work_function_ctx in it. ]*/
+        THANDLE(TIMER) timer_temp = THANDLE_MALLOC(TIMER)(threadpool_timer_dispose);
+        if (timer_temp == NULL)
         {
-            LogError("Failure allocating Timer Instance");
+            LogError("failure in THANDLE_MALLOC(TIMER)(threadpool_timer_dispose=%p)",
+                        threadpool_timer_dispose);
         }
         else
         {
-            timer_instance->work_function = work_function;
-            /* Codes_SRS_THREADPOOL_LINUX_07_057: [ work_function_ctx shall be allowed to be NULL. ]*/
-            timer_instance->work_function_ctx = work_function_ctx;
-            /* Codes_SRS_THREADPOOL_LINUX_45_011: [ threadpool_timer_start shall call interlocked_exchange to set the timer_work_guard to OK_TO_WORK. ]*/
-            (void)interlocked_exchange(&timer_instance->timer_work_guard, OK_TO_WORK);
-            struct sigevent sigev = {0};
-            timer_t time_id = 0;
-
-            sigev.sigev_notify          = SIGEV_THREAD;
-            sigev.sigev_notify_function = on_timer_callback;
-            sigev.sigev_value.sival_ptr = timer_instance;
-
-            /* Codes_SRS_THREADPOOL_LINUX_07_059: [ threadpool_timer_start shall call timer_create and timer_settime to schedule execution. ]*/
-            if (timer_create(CLOCK_REALTIME, &sigev, &time_id) != 0)
+            TIMER * timer_instance = THANDLE_GET_T(TIMER)(timer_temp);
+
+            if (srw_lock_ll_init(&timer_instance->timer_lock) != 0)
             {
-                LogErrorNo("Failure calling timer_create.");
+                LogError("srw_lock_ll_init failed");
             }
```

Comment:
I would expect a Codes for "if any error occurrs then threadpool_timer_start shall fail and return a non zero value.

Reply:
added